Automatically create index name

fisharebest opened the pull request:
<!-- Fill in the relevant information below to help triage your pull request. -->

|      Q       |   A
|------------- | -----------
| Type         | bug

#### Summary

If I add an un-named index to a table using `$table->addIndex(['col1'])`, then an index name is created automatically.
This happens [here](https://github.com/doctrine/dbal/blob/d12da23c3b295911c5e7067321841976e772c01c/src/Schema/Table.php#L150-L154).

But if I add un-named indexes using `Table::__construct`, this does not happen and I get an error.

```
$table = new Table(
            name: 't',
            columns: [
                new Column(name: 'col1', type: new IntegerType()),
                new Column(name: 'col2', type: new IntegerType()),
            ],
            indexes: [
                new Index(name: null, columns: ['col1']),
                new Index(name: null, columns: ['col2']),
            ],
        );
```

`Doctrine\DBAL\Schema\Exception\IndexAlreadyExists: An index with name "" was already defined on table`

This PR copies the same logic for creating index names to the table constructor.

fisharebest committed:
Generate index name

fisharebest commented:
This sets the name of the index (as known to the table), but doesn't set the name of the index (in the index itself).

It can be fixed with `$indexCandidate->_setName($indexName);`, but this doesn't feel like the right solution.

derrabus commented:
Thank you for your PR. Some general notes:

* We do not accept new features for the 4.0.x branch. Our current feature branch is 3.7.x.
* We do not accept changes without tests.

Regarding the proposed change in particular: I don't think that we ever mutate an `Index` object passed to a `Table` and I don't feel comfortable doing that now. Can you elaborate why calling `addIndex()` for that matter does not work for you?

fisharebest commented:
> Can you elaborate why calling addIndex() for that matter does not work for you?

There appears to be two ways to create a table:

1) Creating an empty table object and calling `->addColumn()`, `->addIndex()` …
This works fine.

2) Creating a table object using the constructor (as in my example).
This only works if you choose explicit index names. 
It fails if you want to use the dbal-generated names.

**Should I be able to use the constructor like this?**

If no, then perhaps it could be marked `@internal` to avoid confusion?

> I don't think that we ever mutate an Index object passed to a Table and I don't feel comfortable doing that now.

Agreed - but otherwise we are left with a mismatch between the name of the index (as known to the table) and the name of the index (as reported by itself).  The result is that the index name is not set correctly in the generated SQL.

e.g.

```
var_dump(array_keys($table->getIndexes())); // ["idx_856a5aa8d3ca2f7d", "idx_856a5aa84ac37ec7"]

var_dump($table->getIndex('idx_856a5aa8d3ca2f7d')->getName()); // ""
var_dump($table->getIndex('idx_856a5aa84ac37ec7')->getName()); // ""
```

fisharebest commented:
I've updated the code to use a different approach (does not mutate the `Index` object) and added a test case.

Of course, this depends on the answer to the question "Should I be able to use the constructor like this?"

derrabus commented on src/Schema/Table.php:
```diff
@@ -76,7 +76,15 @@ public function __construct(
-            $this->_addIndex($idx);
+            $idxName = $idx->getName() === '' ? null : $idx->getName();
+
+            if ($idx->isPrimary()) {
+                $this->setPrimaryKey($idx->getColumns(), $idxName);
+            } elseif ($idx->isUnique()) {
+                $this->addUniqueIndex($idx->getColumns(), $idxName, $idx->getOptions());
+            } else {
+                $this->addIndex($idx->getColumns(), $idxName, $idx->getFlags(), $idx->getOptions());
+            }
```
The `Index` class is not final which means that an app might use its own implementation. Your change would break this use-case.

derrabus commented:
> > Can you elaborate why calling `addIndex()` for that matter does not work for you?
> 
> There appears to be two ways to create a table:

Yes, but that does not answer my question.

> **Should I be able to use the constructor like this?**

I don't think that the constructor should mutate (or replace) objects passed to it in any way. All objects should be considered completely configured. With that in mind, I don't see a good way to add the feature of auto-generating index names to the constructor. I'm afraid, this is a dead end.

> If no, then perhaps it could be marked `@internal` to avoid confusion?

No, calling the constructor is fine. But I'm open to suggestions on how this pitfall can be made more obvious.

fisharebest commented:
> I don't think that the constructor should mutate (or replace) objects passed to it in any way.

> But I'm open to suggestions

Perhaps cloning the `Index` before renaming it?

```
foreach ($indexes as $idx) {
    if ($idx->getName() === '') {
        $idx     = clone $idx;
        $idxName = $this->_generateIdentifierName(
            [$this->getName(), ...$idx->getColumns()],
            'idx',
            $this->_getMaxIdentifierLength()
        );
        $idx->_setName($idxName);
    }
    $this->_addIndex($idx);
}
```

derrabus commented:
The more I think about it, the less I believe we should fiddle with the objects passed to the constructor, really. An `Index` instance without a name should not be created in the first place. It's simply not the job of the `Table` class to heal an invalid state of an `Index` object.

fisharebest commented:
> I don't think that the constructor should mutate (or replace) objects passed to it in any way.

This comment appears to suggest otherwise:
https://github.com/doctrine/dbal/blob/d12da23c3b295911c5e7067321841976e772c01c/src/Schema/AbstractAsset.php#L20-L25

derrabus commented:
I'm going to close this PR now because I'm not comfortable merging it. But I want to thank you for your attempt at improving DBAL.

My main problem with the current behavior is the obscure error message that you've mentioned in your PR description. If we can do something about that, I'm open for another PR.

fisharebest commented:
OK - thanks for your time.  I can work around this easily enough in my application code.
